driver-core/src/main/java/com/datastax/driver/core/TabletMap.java

```diff
@@ -0,0 +1,321 @@
+package com.datastax.driver.core;
+
+import com.google.common.annotations.Beta;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Holds currently known tablet mappings. Updated lazily through received custom payloads described
+ * in Scylla's CQL protocol extensions (tablets-routing-v1).
+ */
+@Beta
+public class TabletMap {
+  private static final Logger logger = LoggerFactory.getLogger(TabletMap.class);
+
+  private final Map<KeyspaceTableNamePair, NavigableSet<LazyTablet>> mapping;
+
+  public TabletMap(Map<KeyspaceTableNamePair, NavigableSet<LazyTablet>> mapping) {
+    this.mapping = mapping;
+  }
+
+  public static TabletMap emptyMap() {
+    return new TabletMap(new HashMap<>());
+  }
+
+  /**
+   * Returns the mapping of tables to their tablets.
+   *
+   * @return the Map keyed by (keyspace,table) pairs with Set of tablets as value type.
+   */
+  public Map<KeyspaceTableNamePair, NavigableSet<LazyTablet>> getMapping() {
+    return mapping;
+  }
+
+  /**
+   * Finds hosts that have replicas for a given table and token combination
+   *
+   * @param keyspace the keyspace that table is in
+   * @param table the table name
+   * @param token the token to look for
+   * @return Set of host UUIDS that do have a tablet for given token for a given table.
+   */
+  public Set<UUID> getReplicas(String keyspace, String table, long token) {
+    TabletMap.KeyspaceTableNamePair key = new TabletMap.KeyspaceTableNamePair(keyspace, table);
+
+    if (mapping == null) {
+      logger.trace("This tablets map is null. Returning empty set.");
+      return Collections.emptySet();
+    }
+
+    NavigableSet<LazyTablet> set = mapping.get(key);
+    if (set == null) {
+      logger.trace(
+          "There is no tablets for {}.{} in this mapping. Returning empty set.", keyspace, table);
+      return Collections.emptySet();
+    }
+    LazyTablet row = mapping.get(key).ceiling(LazyTablet.malformedTablet(token));
+    if (row == null || row.firstToken >= token) {
+      logger.trace(
+          "Could not find tablet for {}.{} that owns token {}. Returning empty set.",
+          keyspace,
+          table,
+          token);
+      return Collections.emptySet();
+    }
+
+    HashSet<UUID> uuidSet = new HashSet<>();
+    for (HostShardPair hostShardPair : row.replicas) {
+      uuidSet.add(hostShardPair.getHost());
+    }
+    return uuidSet;
```

Comment:
Like with the shards, this can return replicas that no longer exist. Need to make sure the call sites are prepared for that.